api/pkg/block/service.go

```diff
@@ -43,6 +47,163 @@ func NewAPIService(c client.Client) *APIService {
+	log.Errorf(msg, err)
+	error := response.WriteError(errCode, err)
+	if error != nil {
+		log.Errorf("Response write status and http error failed: %v\n", err)
+		return
+	}
+}
+
+func (s *APIService) checkBackendExists(ctx context.Context, request *restful.Request,
+	response *restful.Response, backendId string) error {
+
+	backendResp, err := s.backendClient.GetBackend(ctx, &backend.GetBackendRequest{Id: backendId})
+	if err != nil {
+		WriteError(response, "Get Backend details failed: %v\n", http.StatusNotFound, err)
+		return err
+	}
+	log.Infof("Backend response = [%+v]\n", backendResp)
+
+	return nil
+}
+
+func (s *APIService) prepareVolumeRequest(request *restful.Request,
+	response *restful.Response) *block.ListVolumeRequest {
+
+	limit, offset, err := common.GetPaginationParam(request)
+	if err != nil {
+		WriteError(response, "Get pagination parameters failed: %v\n", http.StatusInternalServerError, err)
+		return nil
+	}
+
+	sortKeys, sortDirs, err := common.GetSortParam(request)
+	if err != nil {
+		WriteError(response, "Get sort parameters failed: %v\n", http.StatusInternalServerError, err)
+		return nil
+	}
+
+	filterOpts := []string{"name", "type", "status"}
+	filter, err := common.GetFilter(request, filterOpts)
+	if err != nil {
+		WriteError(response, "Get filter failed: %v\n", http.StatusInternalServerError, err)
+		return nil
+	}
+
+	return &block.ListVolumeRequest{
+		Limit:    limit,
+		Offset:   offset,
+		SortKeys: sortKeys,
+		SortDirs: sortDirs,
+		Filter:   filter,
+	}
+}
+
+func (s *APIService) ListVolumes(request *restful.Request, response *restful.Response) {
+	if !policy.Authorize(request, response, "volume:list") {
+		return
+	}
+	log.Info("Received request for Volume List.")
+
+	ctx := common.InitCtxWithAuthInfo(request)
+
+	backendId := request.QueryParameter(common.REQUEST_PATH_BACKEND_ID)
+
+	var listVolumeRequest *block.ListVolumeRequest
+
+	if backendId != "" {
+		//List volumes for the backend.
+		listVolumeRequest = s.listVolumeByBackend(ctx, request, response, backendId)
+	} else {
+		//List all volumes.
+		listVolumeRequest = s.listVolumeDefault(ctx, request, response)
+	}
+
+	if listVolumeRequest == nil {
+		return
+	}
+
+	res, err := s.blockClient.ListVolume(ctx, listVolumeRequest)
+	if err != nil {
+		WriteError(response, "List Volumes failed: %v\n", http.StatusInternalServerError, err)
+		return
+	}
+
+	log.Info("List Volumes successfully.")
+
+	err = response.WriteEntity(res)
+	if err != nil {
+		log.Errorf("Response write entity failed: %v\n", err)
+		return
+	}
+}
+
+func (s *APIService) listVolumeDefault(ctx context.Context, request *restful.Request,
+	response *restful.Response) *block.ListVolumeRequest {
+
+	return s.prepareVolumeRequest(request, response)
+}
+
+func (s *APIService) listVolumeByBackend(ctx context.Context, request *restful.Request, response *restful.Response,
+	backendId string) *block.ListVolumeRequest {
+
+	if s.checkBackendExists(ctx, request, response, backendId) != nil {
```

Comment:
Should it be BadRequest error?

Reply:
Here it is to check for backend resource object existence. It should be a 404 Error if no Backend resource exist